### PR TITLE
[Php80] Add PhpToken polyfill

### PR DESCRIPTION
**Pull request by WinterSilence**
Initial issue #376 
Tests PHP 8

### CODE REVIEW

File: src/Php80/PhpToken.php [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+    /**
+     * @return static[]
+     */
+    public static function tokenize(string $code, int $flags = 0): array
+    {
+        $tokens = [];
+        $line = 1;
+        $position = 0;
+        foreach (\token_get_all($code, $flags) as $token) {
```
Comment (keradus) [inline src/Php80/PhpToken.php:81]: maybe array-map, so we don't need to extend $tokens each iteration?

Comment (WinterSilence) [reply src/Php80/PhpToken.php:81]: @keradus can you explain your idea? you mean `$tokens = \token_get_all($code, $flags);`?

File: src/Php80/Resources/stubs/PhpToken.php [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+if (PHP_VERSION_ID < 80000 && !class_exists('PhpToken') && function_exists('token_get_all')) {
```
Comment (nicolas-grekas) [inline src/Php80/Resources/stubs/PhpToken.php:3]: no need for the class_exists check, and better use extension_loaded which can be optimized by opcache:
`if (PHP_VERSION_ID < 80000 && extension_loaded('tokenizer')) {`

File: src/Php80/PhpToken.php [hunk 3]
```diff
@@ -0,0 +1,110 @@
+<?php
+
+namespace Symfony\Polyfill\Php80;
```
Comment (nicolas-grekas) [inline src/Php80/PhpToken.php:3]: please add the license header

File: composer.json [hunk 4]
```diff
@@ -65,6 +65,9 @@
             "src/Php73/Resources/stubs"
         ]
     },
+    "autoload-dev": {
+        "psr-4": { "Symfony\\Polyfill\\Tests\\": "tests/" }
```
Comment (nicolas-grekas) [inline composer.json:69]: this is not needed

File: src/Php80/PhpToken.php [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+use const T_COMMENT;
+use const T_DOC_COMMENT;
+use const T_OPEN_TAG;
+use const T_WHITESPACE;
+
+/**
+ * Tokenizer: PhpToken polyfill.
+ *
+ * @author Fedonyuk Anton <info@ensostudio.ru>
+ * @link https://www.php.net/manual/en/class.phptoken.php
```
Comment (nicolas-grekas) [inline src/Php80/PhpToken.php:26]: should be removed, we don't add such links in the codebase

File: src/Php80/PhpToken.php [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+use function token_name;
+
+use const T_COMMENT;
+use const T_DOC_COMMENT;
+use const T_OPEN_TAG;
+use const T_WHITESPACE;
+
+/**
+ * Tokenizer: PhpToken polyfill.
```
Comment (nicolas-grekas) [inline src/Php80/PhpToken.php:23]: Can be removed

File: tests/Php80/PhpTokenTest.php [hunk 7]
```diff
@@ -0,0 +1,70 @@
+<?php
+
+namespace Symfony\Polyfill\Tests\Php80;
```
Comment (nicolas-grekas) [inline tests/Php80/PhpTokenTest.php:3]: please add the license header here also

File: tests/Php80/PhpTokenTest.php [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+use PhpToken;
+use Symfony\Polyfill\Php80\PhpToken as PhpTokenPolyfill;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * Tests of PhpToken polyfill, compares polyfill and native instance.
```
Comment (nicolas-grekas) [inline tests/Php80/PhpTokenTest.php:10]: can be removed also

File: tests/Php80/PhpTokenTest.php [hunk 9]
```diff
@@ -0,0 +1,70 @@
+<?php
+
+namespace Symfony\Polyfill\Tests\Php80;
+
+use PhpToken;
```
Comment (nicolas-grekas) [inline tests/Php80/PhpTokenTest.php:5]: should be removed

File: src/Php80/PhpToken.php [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        return $name;
+    }
+
+    /**
+     * @param int|string|array $kind
+     */
+    public function is($kind): bool
+    {
+        $kind = (array) $kind;
+        if (\is_int(\current($kind))) {
```
Comment (nicolas-grekas) [inline src/Php80/PhpToken.php:62]: won't this fail when there are more that one item in $kind and not all of them are integers?
can you also add a test case that covers the `getTokenName` check?

File: tests/Php80/PhpTokenTest.php [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        $polyfillToken = new PhpTokenPolyfill(\T_ECHO, 'echo');
+        $this->assertEquals($token->getTokenName(), $polyfillToken->getTokenName());
+        // single char token
+        $token = new PhpToken(\ord(';'), ';');
+        $polyfillToken = new PhpTokenPolyfill(\ord(';'), ';');
+        $this->assertEquals($token->getTokenName(), $polyfillToken->getTokenName());
+        // unknown token
+        $token = new PhpToken(10000 , "\0");
+        $polyfillToken = new PhpTokenPolyfill(10000 , "\0");
+        $this->assertEquals($token->getTokenName(), $polyfillToken->getTokenName());
```
Comment (nicolas-grekas) [inline tests/Php80/PhpTokenTest.php:44]: let's use assertSame everywhere possible

Comment (WinterSilence) [reply tests/Php80/PhpTokenTest.php:44]: @nicolas-grekas what's reason? it's polyfill class and must be similar

Comment (nicolas-grekas) [reply tests/Php80/PhpTokenTest.php:44]: To ensure that types are also matching

File: src/Php80/PhpToken.php (hunk 10 above)
Comment (WinterSilence) [reply src/Php80/PhpToken.php:62]: @nicolas-grekas yes, need rewrite and test every value

File: tests/Php80/PhpTokenTest.php (hunk 11 above)
Comment (WinterSilence) [reply tests/Php80/PhpTokenTest.php:44]: @nicolas-grekas ok

File: src/Php80/PhpToken.php (hunk 10 above)
Comment (WinterSilence) [reply src/Php80/PhpToken.php]: @nicolas-grekas fixed. In self defense: that such a possibility is undocumented. :)

Comment (WinterSilence) [reply src/Php80/PhpToken.php]: @nicolas-grekas 
 > can you also add a test case that covers the getTokenName check?

removed, [native class don't use it](https://3v4l.org/It16L#v8.1.3)

**Comment by nicolas-grekas** [comment]
Thank you @WinterSilence.

**Comment by WinterSilence** [comment]
@nicolas-grekas thank you too

**Pull request merged**